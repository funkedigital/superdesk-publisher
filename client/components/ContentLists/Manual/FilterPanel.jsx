import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import _ from "lodash";

import MultiSelect from "../../UI/MultiSelect";

class FilterPanel extends React.Component {
  constructor(props) {
    super(props);

    this._isMounted = false;

    this.state = {
      filters: { route: [], author: [] },
      routes: [],
      authors: []
    };
  }

  componentDidMount() {
    this._isMounted = true;

    this.props.publisher.queryRoutes({ type: "collection" }).then(routes => {
      if (this._isMounted) {
        this.setState({ routes }, this.prepareFilters);
      }
    });

    this.loadAuthors();
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  loadAuthors = (page = 1) => {
    this.props.api.users
      .query({
        max_results: 200,
        page: page,
        sort: '[("first_name", 1), ("last_name", 1)]',
        where: {
          is_support: { $ne: true }
        }
      })
      .then(response => {
        let authors = response._items.filter(item => item.is_author);

        if (this._isMounted && authors.length)
          this.setState({ authors: [...this.state.authors, ...authors] });

        if (response._links.next) this.loadAuthors(page + 1);
      });
  };

  handleAuthorChange = arr => {
    let filters = { ...this.state.filters };

    filters.author = arr ? arr : [];
    this.setState({ filters });
  };

  handleRoutesChange = arr => {
    let filters = { ...this.state.filters };

    filters.route = arr ? arr : [];
    this.setState({ filters });
  };

  handleInputChange = (e, a) => {
    let { name, value } = e.target;
    let filters = { ...this.state.filters };

    filters[name] = value;
    this.setState({ filters });
  };

  clear = () => {
    this.setState({ filters: { route: [], author: [] } }, this.save);
  };

  save = () => {
    let filters = _.pickBy({ ...this.state.filters }, _.identity);

    if (this.state.filters.author.length) {
      delete filters.author;
      filters["author[]"] = [];
      this.state.filters.author.map(item => {
        filters["author[]"].push(item.value);
      });
    }

    if (this.state.filters.route.length) {
      delete filters.route;
      filters["route[]"] = [];
      this.state.filters.route.map(item => {
        filters["route[]"].push(item.value);
      });
    }

    this.props.filter(filters);
  };

  render() {
    let routesOptions = [];

    this.state.routes.map(route => {
      routesOptions.push({
        value: parseInt(route.id),
        label: route.name
      });
    });

    let authorsOptions = [];

    this.state.authors.map(author => {
      authorsOptions.push({
        value: author.display_name,
        label: author.display_name
      });
    });

    return (
      <div className="sd-filters-panel sd-filters-panel--border-left">
        <div className="side-panel side-panel--transparent side-panel--shadow-right">
          <div className="side-panel__header side-panel__header--border-b">
            <a
              className="icn-btn side-panel__close"
              sd-tooltip="Close filters"
              flow="left"
              onClick={this.props.toggle}
            >
              <i className="icon-close-small" />
            </a>
            <h3 className="side-panel__heading">Advanced filters</h3>
          </div>
          <div className="side-panel__content">
            <div className="side-panel__content-block">
              <div className="form__row">
                <div className="sd-line-input sd-line-input--no-margin sd-line-input--with-button">
                  <label className="sd-line-input__label">Categories</label>
                  <MultiSelect
                    onSelect={values => this.handleRoutesChange(values)}
                    options={routesOptions}
                    selectedOptions={this.state.filters.route}
                  />
                </div>
              </div>
              <div className="form__row">
                <div className="sd-line-input sd-line-input--no-margin sd-line-input--with-button">
                  <label className="sd-line-input__label">Author</label>
                  <MultiSelect
                    onSelect={values => this.handleAuthorChange(values)}
                    options={authorsOptions}
                    selectedOptions={this.state.filters.author}
                  />
                </div>
              </div>
              <div className="form__row form__row--flex">
                <div className="sd-line-input sd-line-input--no-margin">
                  <label className="sd-line-input__label">Published date</label>
                  <input
                    className="sd-line-input__input"
                    type="date"
                    onChange={this.handleInputChange}
                    name="published_at"
                    value={
                      this.state.filters.published_at
                        ? this.state.filters.published_at
                        : ""
                    }
                  />
                </div>
              </div>
              <div className="form__row form__row--flex">
                <div className="sd-line-input sd-line-input--no-margin">
                  <label className="sd-line-input__label">
                    Published after
                  </label>
                  <input
                    className="sd-line-input__input"
                    type="date"
                    onChange={this.handleInputChange}
                    name="published_after"
                    value={
                      this.state.filters.published_after
                        ? this.state.filters.published_after
                        : ""
                    }
                  />
                </div>
              </div>
              <div className="form__row form__row--flex">
                <div className="sd-line-input sd-line-input--no-margin">
                  <label className="sd-line-input__label">
                    Published before
                  </label>
                  <input
                    className="sd-line-input__input"
                    type="date"
                    onChange={this.handleInputChange}
                    name="published_before"
                    value={
                      this.state.filters.published_before
                        ? this.state.filters.published_before
                        : ""
                    }
                  />
                </div>
              </div>
            </div>
          </div>
          <div className="side-panel__footer side-panel__footer--button-box">
            <div className="flex-grid flex-grid--boxed-small flex-grid--small-2">
              <a className="btn btn--hollow" onClick={this.clear}>
                Clear
              </a>
              <a className="btn btn--primary" onClick={this.save}>
                Filter
              </a>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

FilterPanel.propTypes = {
  filter: PropTypes.func.isRequired,
  toggle: PropTypes.func.isRequired,
  publisher: PropTypes.object.isRequired,
  api: PropTypes.func.isRequired
};

export default FilterPanel;
