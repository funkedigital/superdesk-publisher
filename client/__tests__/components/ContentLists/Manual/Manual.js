import React from "react";
import Manual from "../../../../components/ContentLists/Manual/Manual";
import { render, wait } from "@testing-library/react";
import axios from "axios";

import Publisher from "../../../../__mocks__/publisher";

const publisher = new Publisher();

jest.mock('moment', () => () => ({ fromNow: () => '2 days ago' }));

let lists = [];
publisher.queryLists().then(items => (lists = items));

jest.mock("axios");
jest.mock("react-select", () => props => "div");

let api = () => { };

api.users = {};
api.users.query = function () {
  return new Promise((resolve, reject) => {
    resolve({
      _items: [{ is_author: true, display_name: "test author" }],
      _links: {}
    });
  });
};

const languages = [
  { qcode: 'pl', name: 'Polish' },
  { qcode: 'en', name: 'English' },
  { qcode: 'de', name: 'German' },
]

describe("ContentLists/Manual/Manual", () => {
  it("renders properly", async () => {
    const { container } = render(
      <Manual
        list={lists[0]}
        lists={lists}
        publisher={publisher}
        listEdit={jest.fn()}
        onListUpdate={jest.fn()}
        api={api}
        toggleFilters={jest.fn()}
        isLanguagesEnabled={true}
        languages={languages}
      />
    );

    await wait(() =>
      expect(container.querySelector(".sd-loader")).not.toBeInTheDocument(),
    )
    expect(container.firstChild).toMatchSnapshot();
  });
});
