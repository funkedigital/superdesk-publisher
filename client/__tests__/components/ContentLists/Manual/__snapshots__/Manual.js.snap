// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentLists/Manual/Manual renders properly 1`] = `
<div
  class="flex-grid flex-grid--grow flex-grid--small-2"
>
  <div
    class="flex-grid__item flex-grid__item--d-flex flex-grid__item--column panel-border-right"
  >
    <div
      class="subnav subnav--lower-z-index subnav--dark-blue-grey"
    >
      <button
        class="navbtn navbtn--left"
      >
        <i
          class="icon-th"
        />
      </button>
      <div
        class="flat-searchbar"
        style="background-color: transparent;"
      >
        <div
          class="search-handler"
        >
          <label
            class="trigger-icon"
            for="search-input"
          >
            <i
              class="icon-search"
              style="color: rgb(255, 255, 255);"
            />
          </label>
          <input
            placeholder="Search"
            style="color: rgb(255, 255, 255);"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="dropdown undefined"
      >
        <span>
          <button
            class="dropdown__toggle navbtn navbtn--text-only dropdown-toggle"
          >
            Most shared
            <span
              class="dropdown__caret"
            />
          </button>
        </span>
        <ul
          class="dropdown__menu dropdown__menu--scrollable"
        >
          <li>
            <div
              class="dropdown__menu-label"
            />
          </li>
          <li>
            <button>
              Most shared
            </button>
          </li>
          <li>
            <button>
              Top news
            </button>
          </li>
        </ul>
      </div>
      <div
        class="subnav__stretch-bar"
      />
      <button
        class="btn btn--primary margin--right"
        disabled=""
      >
        Save
      </button>
    </div>
    <div
      class="sd-column-box--3"
    >
      <div
        class="sd-column-box__main-column relative dropZone dropZone--empty"
      >
        <h2
          class="dropZone__heading"
        >
          Drag your Articles here
        </h2>
        <ul
          class="sd-list-item-group sd-shadow--z2"
        />
      </div>
    </div>
  </div>
  <div
    class="flex-grid__item flex-grid__item--d-flex flex-grid__item--column"
  >
    <div
      class="subnav subnav--lower-z-index"
    >
      <button
        class="navbtn navbtn--left navbtn--darker"
        flow="right"
        sd-tooltip="Filter"
      >
        <i
          class="icon-filter-large"
        />
      </button>
      <div
        class="flat-searchbar"
      >
        <div
          class="search-handler"
        >
          <label
            class="trigger-icon"
            for="search-input"
          >
            <i
              class="icon-search"
            />
          </label>
          <input
            placeholder="Search"
            type="text"
            value=""
          />
        </div>
      </div>
      <h3
        class="subnav__page-title"
      >
        All published articles
      </h3>
      <div
        class="subnav__spacer subnav__spacer--no-margin"
      />
      <div
        class="subnav__content-bar sd-flex-no-shrink sd-margin-x--0"
      >
        <div
          class="dropdown dropdown--align-right"
        >
          <span>
            <button
              class="dropdown__toggle navbtn navbtn--text-only dropdown-toggle"
            >
              All Languages
              <span
                class="dropdown__caret"
              />
            </button>
          </span>
          <ul
            class="dropdown__menu dropdown__menu--scrollable"
          >
            <li>
              <button>
                All Languages
              </button>
            </li>
            <li
              class="dropdown__menu-divider"
            />
            <li>
              <button>
                Polish
              </button>
            </li>
            <li>
              <button>
                English
              </button>
            </li>
            <li>
              <button>
                German
              </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="sd-column-box--3"
    >
      <div
        class="sd-column-box__main-column relative"
      >
        <ul
          class="sd-list-item-group sd-shadow--z2"
        >
          <li
            aria-roledescription="Draggable item. Press space bar to lift"
            data-react-beautiful-dnd-drag-handle="0"
            data-react-beautiful-dnd-draggable="0"
            draggable="false"
            tabindex="0"
          >
            <div
              class="sd-list-item"
            >
              <div
                class="sd-list-item__column sd-list-item__column--grow"
              >
                <div
                  class="sd-list-item__row"
                >
                  <span
                    class="sd-overflow-ellipsis sd-list-item__text-strong"
                  >
                    If you let any one have it but me, Basil, I shall never forgive you
                  </span>
                </div>
                <div
                  class="sd-list-item__row"
                >
                  <span
                    class="sd-overflow-ellipsis sd-list-item--element-grow"
                  >
                    2 days ago
                  </span>
                  <span
                    class="label label--success label--hollow"
                  >
                    sport
                  </span>
                </div>
              </div>
            </div>
          </li>
        </ul>
      </div>
      <div
        class="sd-filters-panel sd-filters-panel--border-left"
      >
        <div
          class="side-panel side-panel--transparent side-panel--shadow-right"
        >
          <div
            class="side-panel__header side-panel__header--border-b"
          >
            <a
              class="icn-btn side-panel__close"
              flow="left"
              sd-tooltip="Close filters"
            >
              <i
                class="icon-close-small"
              />
            </a>
            <h3
              class="side-panel__heading"
            >
              Advanced filters
            </h3>
          </div>
          <div
            class="side-panel__content"
          >
            <div
              class="side-panel__content-block"
            >
              <div
                class="form__row"
              >
                <div
                  class="sd-line-input sd-line-input--no-margin sd-line-input--with-button"
                >
                  <label
                    class="sd-line-input__label"
                  >
                    Categories
                  </label>
                  div
                </div>
              </div>
              <div
                class="form__row"
              >
                <div
                  class="sd-line-input sd-line-input--no-margin sd-line-input--with-button"
                >
                  <label
                    class="sd-line-input__label"
                  >
                    Author
                  </label>
                  div
                </div>
              </div>
              <div
                class="form__row form__row--flex"
              >
                <div
                  class="sd-line-input sd-line-input--no-margin"
                >
                  <label
                    class="sd-line-input__label"
                  >
                    Published date
                  </label>
                  <input
                    class="sd-line-input__input"
                    name="published_at"
                    type="date"
                    value=""
                  />
                </div>
              </div>
              <div
                class="form__row form__row--flex"
              >
                <div
                  class="sd-line-input sd-line-input--no-margin"
                >
                  <label
                    class="sd-line-input__label"
                  >
                    Published after
                  </label>
                  <input
                    class="sd-line-input__input"
                    name="published_after"
                    type="date"
                    value=""
                  />
                </div>
              </div>
              <div
                class="form__row form__row--flex"
              >
                <div
                  class="sd-line-input sd-line-input--no-margin"
                >
                  <label
                    class="sd-line-input__label"
                  >
                    Published before
                  </label>
                  <input
                    class="sd-line-input__input"
                    name="published_before"
                    type="date"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="side-panel__footer side-panel__footer--button-box"
          >
            <div
              class="flex-grid flex-grid--boxed-small flex-grid--small-2"
            >
              <a
                class="btn btn--hollow"
              >
                Clear
              </a>
              <a
                class="btn btn--primary"
              >
                Filter
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
